tests/common/snappi_tests/snappi_fixtures.py

```diff
@@ -612,62 +611,69 @@ def tgen_ports(duthost, conn_graph_facts, fanout_graph_facts):  # noqa: F811
         snappi_ports = snappi_fanout_list.get_ports(peer_device=duthost.hostname)
-
         port_speeds = {int(p['speed']) for p in snappi_ports}
         if len(port_speeds) != 1:
             """ All the ports should have the same bandwidth """
             return None
         port_speed = port_speeds.pop()
-        un_ipv4, un_ipv6 = [], []
-        for port in snappi_ports:
-            port['location'] = get_snappi_port_location(port)
+        dutIps = create_ip_list(dut_ip_start, len(snappi_ports), mask=prefix_length)
+        tgenIps = create_ip_list(snappi_ip_start, len(snappi_ports), mask=prefix_length)
+        dutv6Ips = create_ip_list(dut_ipv6_start, len(snappi_ports), mask=v6_prefix_length)
+        tgenv6Ips = create_ip_list(snappi_ipv6_start, len(snappi_ports), mask=v6_prefix_length)
+        for port_id, port in enumerate(snappi_ports):
             port['speed'] = speed_type.get(str(port_speed), port['speed'])
             peer_port = port['peer_port']
             int_addrs = list(config_facts['INTERFACE'][peer_port].keys())
-            port['speed'] = speed_type.get(str(port_speed), port['speed'])
-
-            ipv4_entry = next((a for a in int_addrs if '.' in a), None)
-            if ipv4_entry:
-                port['peer_ip'], port['prefix'] = ipv4_entry.split('/')
-                port['ip'] = get_addrs_in_subnet(ipv4_entry, 2)[1]
-                snappi_ports_all.append(port)
-            else:
-                un_ipv4.append(port)
-
-            ipv6_entry = next((a for a in int_addrs if ':' in a), None)
-            if ipv6_entry:
-                port['peer_ipv6'], port['ipv6_prefix'] = ipv6_entry.split('/')
-                port['ipv6'] = get_ipv6_addrs_in_subnet(ipv6_entry, 2)[1]
-                snappi_ports_all.append(port)
-            else:
-                un_ipv6.append(port)
-
-        if un_ipv4:
-            snappi_ports_all.extend(pre_configure_dut_interface(duthost, un_ipv4, type='ipv4'))
-        if un_ipv6:
-            snappi_ports_all.extend(pre_configure_dut_interface(duthost, un_ipv6, type='ipv6'))
-    return snappi_ports_all
+            for ipver, addr_type in (("ipv4", "IPv4"), ("ipv6", "IPv6")):
+                entry = next((a for a in int_addrs if (":" in a) == (ipver == "ipv6")), None)
+                if ipver == "ipv4":
+                    dut_list, tgen_list, mask = dutIps, tgenIps, prefix_length
+                    peer_ip_key, prefix_key, ip_key = "peer_ip", "prefix", "ip"
+                else:
+                    dut_list, tgen_list, mask = dutv6Ips, tgenv6Ips, v6_prefix_length
+                    peer_ip_key, prefix_key, ip_key = "peer_ipv6", "ipv6_prefix", "ipv6"
+                if entry:
+                    # Already configured on DUT
+                    port[peer_ip_key], port[prefix_key] = entry.split("/")
+                    port[ip_key] = get_addrs_in_subnet(entry, 1, exclude_ips=[entry.split("/")[0]])[0]
```

Comment:
The new approach is cleaner but could select a different IP if there are other devices on the subnet, potentially causing conflicts while calling `get_addrs_in_subnet()`